File: services/api/internal/domain/agent.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

type Agent struct {
	ID           uuid.UUID
	UserID       uuid.UUID
	Name         string
	Description  string
	AgentType    string
	APIKeyHash   string
	Status       string
	LastActiveAt *time.Time
	CreatedAt    time.Time
	UpdatedAt    time.Time
}

type AgentPolicy struct {
	ID          uuid.UUID
	AgentID     uuid.UUID
	PolicyType  string
	Rules       []byte // JSONB
	ScopeFilter []byte // JSONB, nullable
	Priority    int
	Enabled     bool
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

File: packages/shared/src/types/agent.ts
import type { ItemType } from './vault';

export enum AgentType {
  MCPClient = 'mcp_client',
  Claude = 'claude',
  ChatGPT = 'chatgpt',
  Gemini = 'gemini',
  Custom = 'custom',
}

export enum AgentStatus {
  Active = 'active',
  Suspended = 'suspended',
  Revoked = 'revoked',
}

export interface Agent {
  id: string;
  userId: string;
  name: string;
  agentType: AgentType;
  apiKeyHash: string;
  description?: string;
  status: AgentStatus;
  lastAccessAt?: string;
  createdAt: string;
  updatedAt: string;
}

export enum PolicyType {
  ItemScope = 'item_scope',
  ActionPermission = 'action_perm',
  RateLimit = 'rate_limit',
  TimeWindow = 'time_window',
  StepUp = 'step_up',
}

export interface PolicyRules {
  allowedItemTypes?: ItemType[];
  allowedItemIds?: string[];
  allowedFolderIds?: string[];
  deniedItemIds?: string[];

  allowedActions?: ('read' | 'use' | 'proxy')[];

  maxRequests?: number;
  windowSeconds?: number;

  allowedHours?: { start: number; end: number };
  allowedDays?: number[];

  requireApproval?: boolean;
  approvalTimeoutSeconds?: number;
}

export interface AgentPolicy {
  id: string;
  agentId: string;
  policyType: PolicyType;
  rules: PolicyRules;
  scopeFilter?: Record<string, unknown>;
  priority: number;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
}
